### PR TITLE
fix interleaved output in the default panic hook when multiple threads panic simultaneously

### PR DESCRIPTION
**Pull request by jyn514**
previously, we only held a lock for printing the backtrace itself. since all threads were printing to the same file descriptor, that meant random output in the default panic hook from one thread would be interleaved with the backtrace from another. now, we hold the lock for the full duration of the hook, and the output is ordered.

---

i noticed some odd things while working on this you may or may not already be aware of.

- libbacktrace is included as a submodule instead of a normal rustc crate, and as a result uses `cfg(backtrace_in_std)` instead of a more normal `cfg(feature = "rustc-dep-of-std")`. probably this is left over from before rust used a cargo-based build system?
- the default panic handler uses `trace_unsynchronized`, etc, in `sys::backtrace::print`. as a result, the lock only applies to concurrent *panic handlers*, not concurrent *threads*.  in other words, if another, non-panicking, thread tried to print a backtrace at the same time as the panic handler, we may have UB, especially on windows.
    - we have the option of changing backtrace to enable locking when `backtrace_in_std` is set so we can reuse their lock instead of trying to add our own.

### CODE REVIEW

**Comment by workingjubilee** [comment]
> libbacktrace is included as a submodule instead of a normal rustc crate, and as a result uses `cfg(backtrace_in_std)` instead of a more normal `cfg(feature = "rustc-dep-of-std")`. probably this is left over from before rust used a cargo-based build system?

I suspect this is more due to us having needed to internalize, as much as possible, the horrible details of backtraces.

**Comment by workingjubilee** [comment]
I also suspect that in practice, backtraces on Windows are actually much more tame now due to us using `RtlVirtualUnwind`. Am I correct in that, @ChrisDenton?

**Comment by ChrisDenton** [comment]
Yes and no. Symbolization is still a problem area. The APIs we're using are designed to be run from an external "debugger" process and not in-process.

**Comment by jyn514** [comment]
i did find a mutex in windows that looks like it takes care of the multi-threading issues for us: https://github.com/rust-lang/backtrace-rs/blob/72265bea210891ae47bbe6d4f17b493ef0606619/src/dbghelp.rs#L283-L376

**Comment by ChrisDenton** [comment]
Yeah, we need a special process-wide mutex so that it works if multiple Rust DLLs are in the same process (doesn't help if another language or library uses these functions though).

**Comment by workingjubilee** [comment]
> Yeah, we need a special process-wide mutex so that it works if multiple Rust DLLs are in the same process (doesn't help if another language or library uses these functions though).

that sounds fine, tbh. "randomly-included C code" can't coordinate perfectly on this, so we can't either, and our responsibility is discharged by doing the best we can.

File: library/std/src/sys/backtrace.rs [hunk 1]
```diff
@@ -12,13 +12,15 @@ use crate::sync::{Mutex, PoisonError};
...
+pub(crate) fn lock() -> impl Drop {
     static LOCK: Mutex<()> = Mutex::new(());
     LOCK.lock().unwrap_or_else(PoisonError::into_inner)
 }
 
 /// Prints the current backtrace.
-pub fn print(w: &mut dyn Write, format: PrintFmt) -> io::Result<()> {
+///
+/// SAFETY: this function is not Sync. The caller must hold a mutex lock, or there must be only one thread in the program.
+pub(crate) unsafe fn print(w: &mut dyn Write, format: PrintFmt) -> io::Result<()> {
```
Comment (workingjubilee) [inline library/std/src/sys/backtrace.rs:23]: so uh silly question maybe, I guess it's a few more lines of code but why isn't `backtrace::print` a function implemented on the lock that `backtrace::lock` returns, given that we should be holding that specific lock...?

Comment (jyn514) [reply library/std/src/sys/backtrace.rs:23]: no particular reason. i can implement that tomorrow.

File: tests/ui/backtrace/synchronized-panic-handler.run.stderr [hunk 2]
```diff
@@ -0,0 +1,5 @@
+thread '<unnamed>' panicked at $DIR/synchronized-panic-handler.rs:8:5:
+oops oh no woe is me
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+thread '<unnamed>' panicked at $DIR/synchronized-panic-handler.rs:8:5:
+oops oh no woe is me
```
Comment (workingjubilee) [inline tests/ui/backtrace/synchronized-panic-handler.run.stderr:5]: do these wind up interleaved currently on some executions?

I suspect the minimum saturation for threads would have to be... larger, before seeing significant interleaving.

File: library/std/src/sys/backtrace.rs (hunk 1 above)
Comment (workingjubilee) [reply library/std/src/sys/backtrace.rs:23]: and by tomorrow I assume you mean Monday, but of course. :sparkling_heart:

File: tests/ui/backtrace/synchronized-panic-handler.run.stderr (hunk 2 above)
Comment (jyn514) [reply tests/ui/backtrace/synchronized-panic-handler.run.stderr:5]: this is the output from before i made this change: https://github.com/rust-lang/rust/pull/127397/commits/de14f1f932f4f11130be3fed5cd370bd0936032e
you can see it's pretty garbage. after this change i ran the same test ~25 times in a row without failures.

Comment (workingjubilee) [reply tests/ui/backtrace/synchronized-panic-handler.run.stderr:5]: oh wow.

Comment (workingjubilee) [reply tests/ui/backtrace/synchronized-panic-handler.run.stderr:5]: sorry, hadn't noticed there were 2 commits. ^^;

File: library/std/src/sys/backtrace.rs (hunk 1 above)
Comment (jyn514) [reply library/std/src/sys/backtrace.rs]: done!

**Comment by matthiaskrgr** [comment]
@bors r- 
https://github.com/rust-lang/rust/pull/127519#issuecomment-2218148373

**Comment by jyn514** [comment]
lol apparently some test runners set RUST_BACKTRACE=1 on globally? no idea why
anyway i should be able to just unset it for this test

**Comment by jyn514** [comment]
i wonder if it makes sense to label this as `relnotes`.

**Comment by workingjubilee** [comment]
Hmm. Probably. Release team can decide if it fails to meet notability requirements, but it has all the classic details of "something we might want to mention"... error output, locking, and multithreaded cases.

@bors r+

**Comment by jyn514** [comment]
... oops.

should be fixed now.

**Comment by workingjubilee** [comment]
we're getting all sorts of fun discoveries.

**Comment by workingjubilee** [comment]
what is the difference between exec-env and rustc-env...?

@bors r+

**Comment by jyn514** [comment]
> what is the difference between exec-env and rustc-env...?

"the difference between check-pass and run-pass" - one is at compile time and one is when the compiled binary is executed.

**Comment by workingjubilee** [comment]
oh, I see.

**Comment by workingjubilee** [comment]
Ah, I really should have remembered and asked for `//@ needs-threads`

@bors r-

**Comment by jyn514** [comment]
could you mark this as rollup=iffy?

**Comment by workingjubilee** [comment]
yeah.

@bors rollup=iffy r+

**Comment by jyn514** [comment]
congrats to @rust-lang/wg-compiler-performance, this is a really impressively low amount of noise

**Pull request merged**